versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,376 @@
+    collections.put(ValueType.REF, mongoDatabase.getCollection(config.getRefTableName(), InternalRef.class));
+    collections.put(ValueType.VALUE,
+        mongoDatabase.getCollection(config.getValueTableName(), InternalValue.class));
+  }
+
+  /**
+   * Closes the connection this manager creates to a database. If the connection is already closed this method has
+   * no effect.
+   */
+  @Override
+  public void close() {
+    if (null != mongoClient) {
+      mongoClient.close();
+    }
+  }
+
+  @Override
+  public void load(LoadStep loadstep) throws ReferenceNotFoundException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public <V> boolean putIfAbsent(ValueType type, V value) {
+    final MongoCollection<V> collection = (MongoCollection<V>)collections.get(type);
+    if (null == collection) {
+      throw new UnsupportedOperationException(String.format("Unsupported Entity type: %s", type.name()));
+    }
+
+    final UpdateResult result = await(() -> collection.updateOne(
+        Filters.eq(Store.KEY_NAME, ((HasId)value).getId()),
+        new UpdateEntityBson<>(value),
+        UPDATE_UPSERT)).first();
+    return result.getUpsertedId() != null;
+  }
+
+  @Override
+  public <V> void put(ValueType type, V value, Optional<ConditionExpression> conditionUnAliased) {
+    Preconditions.checkArgument(type.getObjectClass().isAssignableFrom(value.getClass()),
+        "ValueType %s doesn't extend expected type %s.", value.getClass().getName(), type.getObjectClass().getName());
+
+    final MongoCollection<V> collection = (MongoCollection<V>)collections.get(type);
+    if (null == collection) {
+      throw new UnsupportedOperationException(String.format("Unsupported Entity type: %s", type.name()));
+    }
+
+    // TODO: Handle ConditionExpressions.
+    await(() -> collection.replaceOne(Filters.eq(Store.KEY_NAME, ((HasId)value).getId()), value, REPLACE_UPSERT));
+  }
+
+  @Override
+  public boolean delete(ValueType type, Id id, Optional<ConditionExpression> condition) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void save(List<SaveOp<?>> ops) {
+    final ListMultimap<MongoCollection<?>, SaveOp<?>> mm = Multimaps.index(ops, l -> collections.get(l.getType()));
+    final ListMultimap<MongoCollection<?>, Object> collectionWrites = Multimaps.transformValues(mm, SaveOp::getValue);
+
+    final List<ObservableSubscriber<InsertManyResult>> subscribers = new ArrayList<>();
+    for (MongoCollection collection : collectionWrites.keySet()) {
+      Lists.partition(collectionWrites.get(collection), paginationSize).forEach(l -> {
+        final ObservableSubscriber<InsertManyResult> subscriber = new ObservableSubscriber<>();
+        subscribers.add(subscriber);
+        collection.insertMany(l, INSERT_UNORDERED).subscribe(subscriber);
+      });
+    }
+
+    // Wait for each of the writes to have completed.
+    subscribers.forEach(s -> {
```

Comment:
Aren't there some built in tools for this? There are definitely for CompletableFuture. Also, I wonder if we should collect all exceptions?

Lastly, the way I read this we keep posting pages even if the first one fails. We should update this so that we cancel future pages if the first one fails.